crates/holochain/src/conductor/conductor/app_connection_auth.rs

```diff
@@ -0,0 +1,243 @@
+use holochain_conductor_api::AppAuthenticationToken;
+use rand::RngCore;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::time::SystemTime;
+
+use crate::conductor::error::{ConductorError, ConductorResult};
+use holochain_types::prelude::InstalledAppId;
+
+pub struct AppAuthTokenStore {
+    issued_tokens: HashMap<AppAuthenticationToken, TokenMeta>,
+}
+
+impl AppAuthTokenStore {
+    pub fn new() -> Self {
+        Self {
+            issued_tokens: HashMap::new(),
+        }
+    }
+
+    /// Issue a token that can be used to authenticate a connection. The token will only be valid
+    /// for use with the specified `installed_app_id` and will expire after `expiry_seconds`.
+    ///
+    /// If `single_use` is true, the token will be invalidated after the first use, successful or not.
+    pub fn issue_token(
+        &mut self,
+        installed_app_id: InstalledAppId,
+        expiry_seconds: u64,
+        single_use: bool,
+    ) -> (AppAuthenticationToken, Option<SystemTime>) {
+        let mut token = [0u8; 64];
+        rand::thread_rng().fill_bytes(&mut token);
+        let token = token.to_vec();
+
+        let expires_at =
+            SystemTime::now().checked_add(std::time::Duration::from_secs(expiry_seconds));
+        self.issued_tokens.insert(
+            token.clone(),
+            TokenMeta {
+                installed_app_id,
+                expires_at,
+                single_use,
+            },
+        );
+        self.remove_expired_tokens();
+
+        (token, expires_at)
+    }
+
+    /// Authenticate a token and return the `InstalledAppId` that the token was issued for.
+    ///
+    /// If `app_id_restriction` is provided, the token will only be valid for the specified `InstalledAppId`.
+    /// This is useful when an app interface is restricted to a single app and tokens that would
+    /// otherwise be valid, are not valid for connecting to this app interface.
+    pub fn authenticate_token(
+        &mut self,
+        token: AppAuthenticationToken,
+        app_id_restriction: Option<InstalledAppId>,
+    ) -> ConductorResult<InstalledAppId> {
+        self.remove_expired_tokens();
+
+        match self.issued_tokens.entry(token) {
+            Entry::Occupied(entry) => {
+                let meta = { entry.get().clone() };
+
+                if meta.single_use {
+                    entry.remove();
+                }
+
+                if let Some(app_id_restriction) = app_id_restriction {
+                    if app_id_restriction != meta.installed_app_id {
+                        return Err(ConductorError::FailedAuthenticationError(
```

Comment:
Should a multiple use token be invalidated if attempted to be used with another app?

Reply:
Not in this case I don't think no. The token hasn't been used to try to get access to another app. The user has just accidentally used it with an app interface that is bound to another application. If a client did this then it probably isn't a recoverable error because there's likely some coding error in selecting app interfaces but I don't see this as a security issue.